src/client/chat.js

```diff
@@ -0,0 +1,538 @@
+      const headerSignature = Buffer.from(packet.headerSignature)
+
+      if (!validateChain(packet, messageSignature, headerSignature, true)) return states.BROKEN_CHAIN
+
+      const valid = client.verifyMessage(client.players[packet.senderUuid].publicKey, packet)
+
+      if (!valid) return states.BROKEN_CHAIN
+
+      client.players[packet.senderUuid].lastSignature = headerSignature
+    } else if (mcData.supportFeature('sessionSignature')) {
+      if (!validateSession(packet)) return states.BROKEN_CHAIN
+
+      const valid = client.verifyMessage(client.players[packet.senderUuid].publicKey, packet, client.players[packet.senderUuid].uuid)
+
+      if (!valid) return states.BROKEN_CHAIN
+
+      client.players[packet.senderUuid].lastSignature = packet.messageSignature
+      client.players[packet.senderUuid].index = packet.index
+    } else {
+      if (!client.verifyMessage(client.players[packet.senderUuid].publicKey, packet)) return states.BROKEN_CHAIN
+    }
+    return states.SECURE
+  }
+
+  function validateHeader (packet) {
+    if (!client.players[packet.senderUuid] || !client.players[packet.senderUuid].publicKey) return states.NOT_SECURE
+
+    const messageSignature = packet.messageSignature ? Buffer.from(packet.messageSignature) : undefined
+    const headerSignature = Buffer.from(packet.headerSignature)
+
+    if (!validateChain(packet, messageSignature, headerSignature, false)) return states.BROKEN_CHAIN
+
+    const verifier = crypto.createVerify('RSA-SHA256')
+    if (messageSignature) verifier.update(messageSignature)
+    verifier.update(concat('UUID', packet.senderUuid, 'buffer', packet.bodyDigest))
+    return verifier.verify(client.players[packet.senderUuid].publicKey, headerSignature) ? states.SECURE : states.BROKEN_CHAIN
+  }
+
+  function validateChain (packet, signature, header, full) {
+    const lastSignature = client.players[packet.senderUuid].lastSignature
+
+    if (!packet.headerSignature || packet.headerSignature.length === 0) return false
+    if (full && !!signature && !!lastSignature && lastSignature.compare(signature) === 0) return true
+    return lastSignature === undefined || lastSignature.compare(header) === 0
+  }
+
+  function validateSession (packet) {
+    const lastSignature = client.players[packet.senderUuid].lastSignature
+    if (!!lastSignature && packet.messageSignature.compare(lastSignature) === 0) return true
+    else return lastSignature === undefined || packet.index > client.players[packet.senderUuid].index
+  }
+
+  function logSystemMessage (packet) {
+    if (!client.chat_log) {
+      client.chat_log = {
+        log: [],
+        old: 0,
+        new: 0,
+        untracked: 0,
+        acknowledgements: []
+      }
+    }
+
+    const id = client.chat_log.new++
+    if (id >= 1024) client.chat_log.old++
+
+    client.chat_log.log[id % 1024] = { type: log.SYSTEM_MESSAGE, packet, timestamp: Date.now() }
```

Comment:
One thing to keep in mind using `id % 1024` is this results in index wrapping. The ordering in chat_log no longer implies leftward for older and rightward for newer, which can make lookups more complex as you need to sort by timestamp.

Reply:
No sorting needs to be done as it tracks index of oldest and newest element. Not pretty, but it works.